myst_nb/parser.py

```diff
@@ -21,6 +27,94 @@ class NotebookParser(MystParser):
     config_section_dependencies = ("parsers",)
 
     def parse(self, inputstring, document):
+
+        # de-serialize the notebook
+        ntbk = nbf.reads(inputstring, nbf.NO_CONVERT)
+
+        # This is a contaner for top level markdown tokens
+        # which we will add to as we walk the document
+        mkdown_tokens = []
+
+        # first we ensure that we are using a 'clean' global context
+        # for parsing, which is setup with the MyST parsing tokens
+        parse_context = ParseContext(
+            find_blocks=SphinxNBRenderer.default_block_tokens,
+            find_spans=SphinxNBRenderer.default_span_tokens,
+        )
+        # TODO hook in the document.reporter here,
+        # to e.g. report on duplicate link/footnote definitions
+        set_parse_context(parse_context)
+
+        for cell_index, nb_cell in enumerate(ntbk.cells):
+
+            # Skip empty cells
+            if len(nb_cell["source"]) == 0:
+                continue
+
+            # skip cells tagged for removal
+            tags = nb_cell.metadata.get("tags", [])
+            if "remove_cell" in tags:
+                continue
+
+            if nb_cell["cell_type"] == "markdown":
+                lines = nb_cell["source"]
+
+                # ensure all lines end with a new line
+                # TODO this is going to be changed in mistletoe, to just be
+                # lines = SourceLines(lines)
+                if isinstance(lines, str):
+                    lines = lines.splitlines(keepends=True)
+                lines = [
+                    line if line.endswith("\n") else "{}\n".format(line)
+                    for line in lines
+                ]
+
+                # parse the markdown;
+                # at this point span/inline level tokens are not yet processed, but
+                # link/footnote definitions are collected/stored in the global context
+                # TODO here it would be ideal to somehow include the cell index
+                # in the `position` attribute of each token
+                # and utilise in within the document.reporter
+                mkdown_tokens.extend(tokenize_block(lines))
+
+                # TODO think of a way to implement the previous
+                # `if "hide_input" in tags:` logic
+
+            elif nb_cell["cell_type"] == "code":
+                # here we do nothing but store the cell as a custom token
+                # TODO here index would instead be part of the position attribute
+                mkdown_tokens.append(NbCodeCell(cell=nb_cell, index=cell_index))
```

Comment:
Could this token keep information about the tags in the cell? And then those tags could be propagated later to deal w/ things like `hide_input`?

I wonder if we could then say "markdown hiding/toggle/etc behavior should be triggered with directives, whereas code cell behavior can be triggered w/ tags"

Reply:
Here we are just storing the entire notebook cell, i.e. 

```json
{
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {"tags": []},
      "source": [],
      "outputs": []
}
```

Reply:
For code cells then, you can utilise the metadata.

For the markdown hiding/toggle/etc behaviour yeh I think the question is what do you actually want to hide? 
- For admonitions (notes, warnings, etc) then certainly just use the `:class:` option, as shown in https://github.com/ExecutableBookProject/sphinx-togglebutton#usage
- For sections, it would be interesting to think about if, at the `SphinxRenderer` level, you want to extract classes from the end of header text. For example, you could write something like

```markdown
# A header with a class {.toggle}
```

The in [render_header](https://github.com/ExecutableBookProject/MyST-Parser/blob/c64f7068ba786a79561adb0017194ebc8bf15ba3/myst_parser/docutils_renderer.py#L299) you would look for these and apply the class to the section (would `sphinx-togglebutton` currently handle this)

Reply:
I was thinking about just implementing a `{toggle}` directive that would wrap everything inside as "toggle-able". The only restriction (I think) would be that it couldn't have a header inside.

Reply:
but I think either way, we don't need to make "toggle entire sections w/ headers" a blocker on this